util/EIP-2124/src/lib.rs

```diff
@@ -0,0 +1,298 @@
+	}
+}
+
+impl rlp::Decodable for ForkHash {
+	fn decode(rlp: &Rlp) -> Result<Self, DecoderError> {
+		rlp.decoder().decode_value(|b| {
+			if b.len() != 4 {
+				return Err(DecoderError::RlpInvalidLength);
+			}
+
+			let mut blob = [0; 4];
+			blob.copy_from_slice(&b[..]);
+
+			return Ok(Self(u32::from_be_bytes(blob)))
+		})
+	}
+}
+
+impl From<H256> for ForkHash {
+	fn from(genesis: H256) -> Self {
+		Self(crc32::checksum_ieee(&genesis[..]))
+	}
+}
+
+impl std::ops::AddAssign<BlockNumber> for ForkHash {
+	fn add_assign(&mut self, height: BlockNumber) {
+		let blob = height.to_be_bytes();
+		self.0 = crc32::update(self.0, &crc32::IEEE_TABLE, &blob)
+	}
+}
+
+impl std::ops::Add<BlockNumber> for ForkHash {
+	type Output = Self;
+	fn add(mut self, height: BlockNumber) -> Self {
+		self += height;
+		self
+	}
+}
+
+/// A fork identifier as defined by EIP-2124.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, RlpEncodable, RlpDecodable)]
+pub struct ForkId {
+	/// CRC32 checksum of the all fork blocks from genesis.
+	pub hash: ForkHash,     
+	/// Next upcoming fork block number, 0 if not yet known.
+	pub next: BlockNumber
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum RejectReason {
+	RemoteStale,
+	LocalIncompatibleOrStale,
+}
+
+#[derive(Clone, Debug)]
+pub struct ForkFilter {
+	pub head: BlockNumber,
+	past_forks: BTreeMap<BlockNumber, ForkHash>,
+	next_forks: BTreeSet<BlockNumber>,
+}
+
+impl ForkFilter {
+	pub fn new<PF, NF>(head: BlockNumber, genesis: H256, past_forks: PF, next_forks: NF) -> Self
+	where
+		PF: IntoIterator<Item = BlockNumber>,
+		NF: IntoIterator<Item = BlockNumber>,
+	{
+		let genesis_fork_hash = ForkHash::from(genesis);
+		Self {
+			head,
+			past_forks: past_forks.into_iter().fold((btreemap! { 0 => genesis_fork_hash }, genesis_fork_hash), |(mut acc, base_hash), block| {
+				let fork_hash = base_hash + block;
+				acc.insert(block, fork_hash);
+				(acc, fork_hash)
+			}).0,
+			next_forks: next_forks.into_iter().collect(),
+		}
+	}
+
+	fn current_fork_hash(&self) -> ForkHash {
+		*self.past_forks.values().next_back().unwrap()
+	}
+
+	fn future_fork_hashes(&self) -> Vec<ForkHash> {
+		self.next_forks.iter().fold((Vec::new(), self.current_fork_hash()), |(mut acc, hash), fork| {
+			let next = hash + *fork;
+			acc.push(next);
+			(acc, next)
+		}).0
+	}
+
+	pub fn insert_past_fork(&mut self, height: BlockNumber) {
+		self.past_forks.insert(height, self.current_fork_hash() + height);
+	}
+
+	pub fn insert_next_fork(&mut self, height: BlockNumber) {
+		self.next_forks.insert(height);
+	}
+
+	pub fn promote_next_fork(&mut self, height: BlockNumber) -> bool {
+		let promoted = self.next_forks.remove(&height);
+		if promoted {
+			self.insert_past_fork(height);
+		}
+		promoted
+	}
+
+	pub fn is_valid(&self, fork_id: ForkId) -> Result<(), RejectReason> {
+		// 1) If local and remote FORK_HASH matches...
+		if self.current_fork_hash() == fork_id.hash {
+			if fork_id.next == 0 {
+				// 1b
```

Comment:
Let's add a clear reference to what case `1b` refers to.

Reply:
done